File: common/src/main/java/me/kvalbrus/multibans/common/punishment/MultiTemporaryPunishment.java
package me.kvalbrus.multibans.common.punishment;

import java.util.List;
import me.kvalbrus.multibans.api.punishment.creator.PunishmentCreator;
import me.kvalbrus.multibans.api.punishment.PunishmentStatus;
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget;
import me.kvalbrus.multibans.api.punishment.PunishmentType;
import me.kvalbrus.multibans.api.punishment.TemporaryPunishment;
import me.kvalbrus.multibans.common.managers.PluginManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class MultiTemporaryPunishment extends MultiPunishment
    implements TemporaryPunishment {

    private long startedDate;

    private PunishmentCreator cancellationCreator;

    private long cancellationDate;

    private String cancellationReason;

    private boolean cancelled;

    private long duration;

    public MultiTemporaryPunishment(@NotNull PluginManager pluginManager,
                                    @NotNull PunishmentType type,
                                    @NotNull String id,
                                    @NotNull PunishmentTarget target,
                                    @NotNull PunishmentCreator creator,
                                    long createdDate,
                                    long startedDate,
                                    long duration,
                                    @Nullable String reason,
                                    @Nullable String comment,
                                    @Nullable PunishmentCreator cancellationCreator,
                                    long cancellationDate,
                                    @Nullable String cancellationReason,
                                    @NotNull List<String> servers,
                                    boolean cancelled) {
        super(pluginManager, type, id, target, creator, createdDate, reason, comment, servers);
        this.startedDate = startedDate;
        this.cancellationCreator = cancellationCreator;
        this.cancellationDate = cancellationDate;
        this.cancellationReason = cancellationReason;
        this.cancelled = cancelled;
        this.duration = duration;
    }

    @Override
    public synchronized void activate() {
        this.cancelled = false;

        List<? extends MultiTemporaryPunishment> activePunishments = this.getPluginManager()
            .getPunishmentManager().getActivePunishments(this.getTargetUniqueId(), this.getClass());

        final int size = activePunishments.size();
        if (size > 0) {
            activePunishments.sort(null);

            MultiTemporaryPunishment last = activePunishments.get(size - 1);

            this.startedDate = last.startedDate + last.duration;

        } else {
            this.startedDate = this.getCreatedDate();
        }

        super.activate();
    }

    @Override
    public synchronized void deactivate() {
        this.deactivate(null, System.currentTimeMillis(), null);
    }

    @Override
    public void deactivate(@Nullable PunishmentCreator cancellationCreator, long cancellationDate) {
        this.deactivate(cancellationCreator, cancellationDate, null);
    }

    @Override
    public void deactivate(@Nullable PunishmentCreator cancellationCreator,
                           long cancellationDate,
                           @Nullable String cancellationReason) {
        if (this.cancelled) {
            return;
        }

        List<? extends MultiTemporaryPunishment> history = this.getPluginManager()
            .getPunishmentManager().getPlayerHistory(this.getTarget().getUniqueId(), this.getClass());

        if(history.stream().noneMatch(punishment -> punishment.getId().equals(this.getId()))) {
            return;
        }

        history.removeIf(MultiTemporaryPunishment::isCancelled);
        history.sort(null);

        int size = history.size();
        if (size > 0) {
            MultiTemporaryPunishment prev = null, curr = null;
            int index = -1;
            for (MultiTemporaryPunishment punishment : history) {
                if (punishment.getId().equals(this.getId())) {
                    index = history.indexOf(punishment);
                    break;
                }
            }

            if (index == 0) {
                while (index < size - 1) {
                    curr = history.get(index + 1);
                    if (prev == null) {
                        curr.startedDate = history.get(0).getStartedDate();
                    } else {
                        curr.startedDate = prev.startedDate + prev.duration;
                    }

                    prev = curr;
                    index++;

                    curr.updateData();
                }
            } else if (index != size - 1) {
                prev = history.get(index - 1);
                while (index < size - 1) {
                    curr = history.get(index + 1);
                    curr.startedDate = prev.startedDate + prev.duration;

                    prev = curr;
                    index++;

                    curr.updateData();
                }
            }
        }

        this.cancelled = true;
        this.cancellationCreator = cancellationCreator;
        this.cancellationDate = cancellationDate;
        this.cancellationReason = cancellationReason;

        this.getPluginManager().deactivatePunishment(this);
        this.updateData();
    }

    @Override
    public synchronized void delete() {
        this.deactivate();
        this.deleteData();
    }

    @NotNull
    @Override
    public PunishmentStatus getStatus() {
        if (this.cancelled) {
            return PunishmentStatus.CANCELLED;
        } else if (System.currentTimeMillis() > this.getCreatedDate() + this.duration) {
            return PunishmentStatus.PASSED;
        } else {
            return PunishmentStatus.ACTIVE;
        }
    }

    @Nullable
    @Override
    public final PunishmentCreator getCancellationCreator() {
        return this.cancellationCreator;
    }

    @Override
    public final long getCancellationDate() {
        return this.cancellationDate;
    }

    @Nullable
    @Override
    public final String getCancellationReason() {
        return this.cancellationReason;
    }

    @Override
    public final boolean isCancelled() {
        return this.cancelled;
    }

    @Override
    public final long getDuration() {
        return this.duration;
    }

    @Override
    public final long getStartedDate() {
        return this.startedDate;
    }

    @Override
    public final synchronized void setCancellationDate(long cancellationDate) {
        this.cancellationDate = cancellationDate;
        this.updateData();
    }

    @Override
    public final synchronized void setCancellationReason(@Nullable String cancellationReason) {
        this.cancellationReason = cancellationReason;
        this.updateData();
    }

    @Override
    public final synchronized void setCancelled(boolean cancelled) {
        this.cancelled = cancelled;
        this.updateData();
    }

    @Override
    public void setCancellationCreator(@Nullable PunishmentCreator cancellationCreator) {
        this.cancellationCreator = cancellationCreator;
        this.updateData();
    }

    @Override
    public final synchronized void setDuration(long duration) {
        this.duration = duration;
        this.updateData();
    }

    @Override
    public final synchronized void setStartedDate(long startedDate) {
        this.startedDate = startedDate;
        this.updateData();
    }
}

File: common/src/main/kotlin/me/kvalbrus/multibans/common/punishment/MultiTemporaryPunishment.kt
package me.kvalbrus.multibans.common.punishment

import me.kvalbrus.multibans.api.punishment.punishments.PunishmentStatus
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentType
import me.kvalbrus.multibans.api.punishment.TemporaryPunishment
import me.kvalbrus.multibans.api.punishment.action.Action
import me.kvalbrus.multibans.api.punishment.action.ActivationAction
import me.kvalbrus.multibans.api.punishment.action.CreationAction
import me.kvalbrus.multibans.api.punishment.action.DeactivationAction
import me.kvalbrus.multibans.api.punishment.executor.PunishmentExecutor
import me.kvalbrus.multibans.common.managers.PluginManager
import me.kvalbrus.multibans.common.punishment.action.MultiActivationAction
import me.kvalbrus.multibans.common.punishment.action.MultiDeactivationAction

abstract class MultiTemporaryPunishment : MultiPunishment, TemporaryPunishment {

    private var _activations: MutableList<Action> = mutableListOf()
    private var _deactivations: MutableList<Action> = mutableListOf()
    private var _startedDate: Long
    private var _duration: Long
    private var _cancelled: Boolean

    constructor(
        pluginManager: PluginManager,
        type: PunishmentType,
        id: String,
        creationAction: CreationAction,
        startedDate: Long,
        duration: Long,
        comment: String,
        servers: List<String>,
        cancelled: Boolean) : super(pluginManager, type, id, creationAction, comment, servers) {
        this._startedDate = startedDate
        this._duration = duration
        this._cancelled = cancelled
    }

    constructor(
        pluginManager: PluginManager,
        type: PunishmentType,
        id: String,
        creationAction: CreationAction,
        activations: MutableList<Action>,
        deactivations: MutableList<Action>,
        startedDate: Long,
        duration: Long,
        comment: String,
        servers: List<String>,
        cancelled: Boolean) : this(pluginManager, type, id, creationAction, startedDate, duration, comment, servers, cancelled) {
        this._activations = activations
        this._deactivations = deactivations
    }

    final override val activations: List<Action>
        get() = ArrayList(this._activations)

    final override val deactivations: List<Action>
        get() = ArrayList(this._deactivations)

    final override val cancelled: Boolean
        get() = this._cancelled

    final override val duration: Long
        get() = this._duration

    final override val startedDate: Long
        get() = this._startedDate

    @Synchronized
    override fun create() {
        val activePunishments = pluginManager
            .punishmentManager.getActivePunishments(this.target.uniqueId, this.javaClass)
        val size = activePunishments.size
        if (size > 0) {
            activePunishments.sorted()
            val last = activePunishments[size - 1]
            this._startedDate = last.startedDate + last.duration
        } else {
            this._startedDate = this.createdDate
        }

        super.create()
    }

    @Synchronized
    override fun activate(executor: PunishmentExecutor, date: Long, reason: String) : Boolean {
        return this.activate(MultiActivationAction(this.id, this._activations.size + 1, executor, date, reason))
    }

    @Synchronized
    override fun activate(action: ActivationAction) : Boolean {
        if (this._cancelled) {
            this._cancelled = false

            this._activations.add(action)

            val activePunishments = pluginManager
                .punishmentManager.getActivePunishments(this.target.uniqueId, this.javaClass)
            val size = activePunishments.size
            if (size > 0) {
                activePunishments.sorted()
                val last = activePunishments[size - 1]
                this._startedDate = last.startedDate + last.duration
            } else {
                this._startedDate = this.createdDate
            }

            this.updateData()
            this.pluginManager.activatePunishment(this, action)
            this.sendMessageAboutActivate()

            return true
        } else {
            return false
        }
    }

    @Synchronized
    override fun deactivate(executor: PunishmentExecutor, date: Long, reason: String): Boolean {
        return this.deactivate(MultiDeactivationAction(this.id, this._deactivations.size + 1, executor, date, reason))
    }

    @Synchronized
    override fun deactivate(action: DeactivationAction) : Boolean {
        if (!this._cancelled) {
            this.deleteAndSortPunishments()

            this._cancelled = true

            this._deactivations.add(action)
            this.updateData()
            this.pluginManager.deactivatePunishment(this, action)
            this.sendMessageAboutDeactivate()

            return true
        } else {
            return false
        }
    }

    @Synchronized
    override fun delete() {
        if (!this._cancelled) {
            this.deleteAndSortPunishments()
        }

        super.delete()
    }

    override fun getStatus(): PunishmentStatus {
        return if (this.cancelled) {
            PunishmentStatus.CANCELLED
        } else if (System.currentTimeMillis() > createdDate + duration) {
            PunishmentStatus.PASSED
        } else {
            PunishmentStatus.ACTIVE
        }
    }

    override fun lastActivator(): PunishmentExecutor {
        if (this.activations.isEmpty()) {
            return this.creationAction.executor
        } else {
            var maxTimeIndex = 0
            val activations = this.activations
            for (i in 0..activations.size) {
                if (activations[maxTimeIndex].date < activations[i].date) {
                    maxTimeIndex = i
                }
            }

            return activations[maxTimeIndex].executor
        }
    }

    override fun lastDeactivator(): PunishmentExecutor? {
        if (this.deactivations.isEmpty()) {
            return null
        } else {
            var maxTimeIndex = 0
            val deactivations = this.deactivations
            for (i in 0..deactivations.size) {
                if (deactivations[maxTimeIndex].date < deactivations[i].date) {
                    maxTimeIndex = i
                }
            }

            return deactivations[maxTimeIndex].executor
        }
    }

    private fun sendMessageAboutActivate() {
        sendMessageToListeners(this.activateMessageForListener)
        sendMessageToCreator(this.activateMessageForExecutor)
        sendMessageToTarget(this.activateMessageForTarget)
    }

    private fun sendMessageAboutDeactivate() {
        sendMessageToListeners(this.deactivateMessageForListener)
        sendMessageToCreator(this.deactivateMessageForExecutor)
        sendMessageToTarget(this.deactivateMessageForTarget)
    }

    abstract val activateMessageForListener: String
    abstract val activateMessageForExecutor: String
    abstract val activateMessageForTarget: String

    abstract val deactivateMessageForListener: String
    abstract val deactivateMessageForExecutor: String
    abstract val deactivateMessageForTarget: String

    @Synchronized
    private fun deleteAndSortPunishments() {
        val history = ArrayList<MultiTemporaryPunishment>()
        for (pun in pluginManager.punishmentManager.getPlayerHistory(target.uniqueId, this.javaClass)) {
            if (pun is MultiTemporaryPunishment) {
                history.add(pun)
            }
        }

        if (history.stream().noneMatch { punishment: MultiTemporaryPunishment -> punishment.id == id }) {
            return
        }

        history.removeIf { obj: MultiTemporaryPunishment -> obj.cancelled }
        history.sort()
        val size = history.size
        if (size > 0) {
            var prev: MultiTemporaryPunishment? = null
            var curr: MultiTemporaryPunishment?
            var index = -1
            for (punishment in history) {
                if (punishment.id == id) {
                    index = history.indexOf(punishment)
                    break
                }
            }

            if (index == 0) {
                while (index < size - 1) {
                    curr = history[index + 1]
                    if (prev == null) {
                        curr._startedDate = history[0].startedDate
                    } else {
                        curr._startedDate = prev.startedDate + prev.duration
                    }
                    prev = curr
                    index++
                    curr.updateData()
                }
            } else if (index != size - 1) {
                prev = history[index - 1]
                while (index < size - 1) {
                    curr = history[index + 1]
                    curr._startedDate = prev!!.startedDate + prev.duration
                    prev = curr
                    index++
                    curr.updateData()
                }
            }
        }
    }
}